### PR TITLE
Allow `BasisStatePreparation` to be JIT compiled with JAX

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
Previously the `BasisStatePreparation` template cannot be JIT compiled with JAX. See the GitHub issue below for more information.

**Description of the Change:**
Allow `BasisStatePreparation` to be JIT compiled.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
Closes https://github.com/PennyLaneAI/pennylane/issues/3225
Closes https://github.com/PennyLaneAI/pennylane/issues/3224

### CODE REVIEW

File: pennylane/templates/embeddings/basis.py [hunk 1]
```diff
@@ -125,9 +135,4 @@ def compute_decomposition(wires, basis_state):  # pylint: disable=arguments-diff
         [PauliX(wires=['a']),
          PauliX(wires=['c'])]
         """
-        ops_list = []
-        for wire, bit in zip(wires, basis_state):
-            if bit == 1:
-                ops_list.append(qml.PauliX(wire))
-
-        return ops_list
+        return qml.BasisStatePreparation.compute_decomposition(basis_state, wires)
```
Comment (rmoyard) [inline pennylane/templates/embeddings/basis.py:138]: Great change here!

File: pennylane/templates/state_preparations/basis.py [hunk 2]
```diff
@@ -103,8 +108,21 @@ def compute_decomposition(basis_state, wires):  # pylint: disable=arguments-diff
         [PauliX(wires=['a']),
         PauliX(wires=['b'])]
         """
-        op_list = []
-        for wire, state in zip(wires, basis_state):
-            if state == 1:
-                op_list.append(qml.PauliX(wire))
+        # if the state is not a JAX tracer or TensorFlow non-eager tensor, compute
+        # the decomposition using PauliX gates on the '1' wires
+        if not qml.math.is_abstract(basis_state):
```
Comment (rmoyard) [inline pennylane/templates/state_preparations/basis.py:113]: Why the decomposition depends on the type of the basis_state? I guess for tracers it is because you don't want to give parameters, but you we don't use this one for both cases?

File: tests/templates/test_embeddings/test_basis.py [hunk 3]
```diff
@@ -203,7 +207,32 @@ def test_jax(self, tol):
...
+        features = jnp.array([0, 1, 0])
+
+        dev = qml.device("default.qubit", wires=3)
+
+        circuit = qml.QNode(circuit_template, dev, interface="jax")
+        circuit2 = qml.QNode(circuit_decomposed, dev, interface="jax")
+
+        res = circuit(features)
+        res2 = circuit2(features)
+        assert qml.math.allclose(res, res2, atol=tol, rtol=0)
```
Comment (rmoyard) [inline tests/templates/test_embeddings/test_basis.py:231]: Minor: Add a test fot tensorflow autograph

File: pennylane/templates/embeddings/basis.py [hunk 4]
```diff
@@ -72,9 +73,16 @@ def circuit(feature_vector):
...
-        if isinstance(features, int):
-            bin_string = f"{features:b}".zfill(len(wires))
-            features = [1 if d == "1" else 0 for d in bin_string]
+        if isinstance(features, list):
+            features = qml.math.stack(features)
+
+        if qml.math.shape(features) == ():
+            if not qml.math.is_abstract(features) and features >= 2 ** len(wires):
+                raise ValueError(
+                    f"Features must be of length {len(wires)}, got features={features} which is >= {2 ** len(wires)}"
```
Comment (rmoyard) [inline pennylane/templates/embeddings/basis.py:82]: That's a new error message if I am not mistaken but I did not find a corresponding test?

**Review by rmoyard — COMMENTED** [review]
Great fix @eddddddy 💯 I've some questions and suggestions. Happy to approve after that!

File: pennylane/templates/state_preparations/basis.py (hunk 2 above)
Comment (eddddddy) [reply pennylane/templates/state_preparations/basis.py:113]: I'm reluctant to change existing behaviour if there's nothing wrong with it. And also because there will be unnecessary operators in cases where the basis state is close to being all zeros.

For tracers, since we don't know if the value of the basis state is 0 or 1, we need these extra operators just in case a 1 is passed in.

File: pennylane/templates/embeddings/basis.py (hunk 4 above)
Comment (eddddddy) [reply pennylane/templates/embeddings/basis.py:82]: Previously this was the same case as lines 95-97, and was tested in the `TestInputs` class. I had to split this off into two cases: one where `features` is an integer and the other where `features` is a vector, as these are fundamentally different. I kept the same test though.

File: pennylane/templates/embeddings/basis.py [hunk 5]
```diff
@@ -88,10 +96,12 @@ def __init__(self, features, wires, do_queue=True, id=None):
                 f"Features must be of length {len(wires)}; got length {n_features} (features={features})."
             )
 
-        features = list(qml.math.toarray(features))
-
-        if not set(features).issubset({0, 1}):
-            raise ValueError(f"Basis state must only consist of 0s and 1s; got {features}")
+        # don't perform the validation if the features is a JAX tracer or a TensorFlow
+        # non-eager tensor
+        if not qml.math.is_abstract(features):
```
Comment (antalszava) [inline pennylane/templates/embeddings/basis.py:101]: It looks like we might evaluate `not qml.math.is_abstract(features)` twice in certain cases in this method. How about assigning it to an object earlier on and using that object at the two occurrences?

Comment (eddddddy) [reply pennylane/templates/embeddings/basis.py]: Fair, updated.

File: tests/templates/test_state_preparations/test_basis_state_prep.py [hunk 6]
```diff
@@ -76,6 +76,34 @@ def circuit():
 
         assert np.allclose(output_state, target_state, atol=tol, rtol=0)
 
+    @pytest.mark.jax
+    @pytest.mark.parametrize(
+        "basis_state,wires,target_state",
+        [
+            ([0, 1], [0, 1], [0, 1, 0]),
+            ([1, 1, 0], [0, 1, 2], [1, 1, 0]),
```
Comment (antalszava) [inline tests/templates/test_state_preparations/test_basis_state_prep.py:84]: Let's add some non-std wires too, like randomly ordered wires, wires of custom wire labels, etc.

**Review by antalszava — CHANGES_REQUESTED** [review]
Great fix! :1st_place_medal:

Don't forget to add entries to the changelog for both templates!

File: pennylane/templates/embeddings/basis.py [hunk 7]
```diff
@@ -88,10 +96,12 @@ def __init__(self, features, wires, do_queue=True, id=None):
                 f"Features must be of length {len(wires)}; got length {n_features} (features={features})."
             )
 
-        features = list(qml.math.toarray(features))
-
-        if not set(features).issubset({0, 1}):
-            raise ValueError(f"Basis state must only consist of 0s and 1s; got {features}")
+        # don't perform the validation if the features is a JAX tracer or a TensorFlow
+        # non-eager tensor
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/templates/embeddings/basis.py]: Removed the comment

File: pennylane/templates/state_preparations/basis.py [hunk 8]
```diff
@@ -72,10 +74,13 @@ def __init__(self, basis_state, wires, do_queue=True, id=None):
...
                 )
 
-            if not all(bit in [0, 1] for bit in state):
-                raise ValueError(
-                    f"Basis states must only consist of 0s and 1s; state {i} is {state}"
-                )
+            # don't perform the validation if the state is a JAX tracer or a TensorFlow
+            # non-eager tensor
+            if not qml.math.is_abstract(state):
+                if not all(bit in [0, 1] for bit in state):
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/templates/state_preparations/basis.py:80]: Making this change here since it's simple enough.

File: tests/templates/test_embeddings/test_basis.py [hunk 9]
```diff
@@ -171,6 +170,9 @@ def test_list_and_tuples(self, tol):
         res2 = circuit2(tuple(features))
         assert qml.math.allclose(res, res2, atol=tol, rtol=0)
 
+        res = circuit(2)
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply tests/templates/test_embeddings/test_basis.py:173]: `circuit2` is just the decomposed circuit (i.e. the expected circuit). The expected result is the same as the one above, hence why I didn't execute again.

File: pennylane/templates/embeddings/basis.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/templates/embeddings/basis.py]: Reverted 👍

File: pennylane/templates/state_preparations/basis.py [hunk 10]
```diff
@@ -103,8 +108,21 @@ def compute_decomposition(basis_state, wires):  # pylint: disable=arguments-diff
...
+            op_list = []
+            for wire, state in zip(wires, basis_state):
+                if state == 1:
+                    op_list.append(qml.PauliX(wire))
+            return op_list
+
+        # otherwise, use RX gates with 0 or pi angles
+        op_list = [
+            qml.PhaseShift(state * np.pi / 2, wire) for wire, state in zip(wires, basis_state)
+        ]
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/templates/state_preparations/basis.py]: Not sure if there's a difference in behaviour in these two. I did change it to the second option though because I agree it looks cleaner.

File: tests/templates/test_state_preparations/test_basis_state_prep.py (hunk 6 above)
Comment (eddddddy) [reply tests/templates/test_state_preparations/test_basis_state_prep.py:84]: I added a case for randomly-ordered wires. The use of custom wire labels isn't "allowed" by these tests because we use the `qubit_device_3_wires` fixture.

**Review by rmoyard — APPROVED** [review]
Thanks @eddddddy it looks good on my side!

**Review by timmysilv — APPROVED** [review]
nice addition!

File: tests/templates/test_embeddings/test_basis.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/templates/test_embeddings/test_basis.py:173]: SG, thanks!

**Review by antalszava — APPROVED** [review]
Looks good! :100:
